### PR TITLE
[scudo] Update header without read-modify-write operation

### PR DESCRIPTION
**Pull request by ChiaHungDuan**
We used to update the deallocated block with
atomic_compare_exchange_strong to ensure the concurrent double-free will be detected. However, this operation incurs huge performance overhead which takes over 50% execution time in deallocate(). Given that we already have the checksum to guard the most double-free cases and other block verifications in the primary allocator, use atomic-store instead.

### CODE REVIEW

**Review by cferris1000 — APPROVED** [review]
LGTM.

**Comment by cryptoad** [comment]
This is risky.

This is a race detection mechanism where we detect if 2 chunks are modified concurrently. There is a window of time where 2 threads can free the same chunk, both headers will be read, pass the check and written as a free chunk header. If we do not detect this, then the free chunk will be added to 2 caches, and further badness will ensue.

I agree this is an expensive check, and I initially considered both options (cmpxchg vs read-store). If you look for "race on chunk header at address" in Google, you'll find stuff like https://github.com/Instabug/Instabug-Android/issues/435. So those races do happen in the wild, and while it would be complicated to exploit, I wouldn't think it to be impossible.

If you do want to remove this - at the cost of security - you might want to remove reportHeaderRace as well though.

**Comment by ChiaHungDuan** [comment]
> This is risky.
> 
> This is a race detection mechanism where we detect if 2 chunks are modified concurrently. There is a window of time where 2 threads can free the same chunk, both headers will be read, pass the check and written as a free chunk header. If we do not detect this, then the free chunk will be added to 2 caches, and further badness will ensue.
> 
> I agree this is an expensive check, and I initially considered both options (cmpxchg vs read-store). If you look for "race on chunk header at address" in Google, you'll find stuff like [Instabug/Instabug-Android#435](https://github.com/Instabug/Instabug-Android/issues/435). So those races do happen in the wild, and while it would be complicated to exploit, I wouldn't think it to be impossible.
> 
> If you do want to remove this - at the cost of security - you might want to remove reportHeaderRace as well though.

Thanks, we really want to get your input.

We have measured cmpxchg, xchg and atomic-store. The former two show the same performance overhead (especially on platforms with more cores). We also simulated the behavior and the success rate is about 1 success from 2 millions attempts on x86_64 machine with 96 cores. Like you said, it's not impossible and seems not long to hit the case. 

I tried to study some double-free attacks and they usually try to override the metadata in the freelist. In Scudo, the way it stores the freelist is quite different from those cases. That gives me the thought that maybe the trade-off is acceptable. Besides, attacking on this may have lower value than compromise the checksum. When the checksum is compromised, then forfeit a double-free case will be easier. Even though, I'm still uncertain.

Another thought is to slightly harden the double-free protection. We can check the header status at allocation (still with atomic-load/atomic-store) as well. Even though we may only be able to check if it's allocated, but the benefit of this is that we rely on the mutex while accessing primary allocator and don't need to introduce additional synchronization. We still miss the case of accessing cache only but I guess it's a light enhancement. 

What do you think?

**Comment by cryptoad** [comment]
> What do you think?

I mean, it's fair. If the performance gains are significant, the protection might not be worth it. I initially went the side of security. I agree that the window for the race makes it difficult, and failure to win the race would mean an abort() which would prevent further attempts, or at least make them very noisy.

You might want to run that through the Fuchsia people as well, make sure that everyone knows it's coming. I do think that the race will happen in the wild (another example @ https://github.com/tensorflow/tensorflow/issues/58143), and will end up manifesting in some weird way that will be hard to debug - which is probably not worth the performance cost either.

I'll leave it to you and Christopher to make the decision.

File: compiler-rt/lib/scudo/standalone/chunk.h [hunk 1]
```diff
@@ -128,19 +128,6 @@ inline void loadHeader(u32 Cookie, const void *Ptr,
...
-                                  UnpackedHeader *NewUnpackedHeader,
-                                  UnpackedHeader *OldUnpackedHeader) {
-  NewUnpackedHeader->Checksum =
-      computeHeaderChecksum(Cookie, Ptr, NewUnpackedHeader);
-  PackedHeader NewPackedHeader = bit_cast<PackedHeader>(*NewUnpackedHeader);
-  PackedHeader OldPackedHeader = bit_cast<PackedHeader>(*OldUnpackedHeader);
-  if (UNLIKELY(!atomic_compare_exchange_strong(
-          getAtomicHeader(Ptr), &OldPackedHeader, NewPackedHeader,
-          memory_order_relaxed)))
-    reportHeaderRace(Ptr);
```
Comment (cryptoad) [inline compiler-rt/lib/scudo/standalone/chunk.h:141]: Remove `reportHeaderRace` and associated constants & code.

**Review by cryptoad — CHANGES_REQUESTED** [review]

**Comment by cferris1000** [comment]
Yeah, we agree there is a small loss of security. We are trying to think of another method to detect this error that might not give as much information, but prevent the error from occurring. For example, if this happened then it's likely the same allocation winds up twice in a list, and maybe there is a way we can modify the header when doing an allocation that prevents the same allocation from being used twice. If you've got any ideas along that line, let us know.

Our thinking is that as a pure bug, the chance of this happening is low. If you trigger this problem, you will likely trigger a normal double free so the chance of only every seeing the multiple threads double freeing at the same time is low. Thus you will still see the double free error.

File: compiler-rt/lib/scudo/standalone/chunk.h (hunk 1 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/chunk.h:141]: Done

**Review by cryptoad — APPROVED** [review]

**Comment by ChiaHungDuan** [comment]
Also created an internal bug to bring some enhancements back later

**Pull request merged**